### ISSUE TITLE
Initial values for any type

### ISSUE THREAD

**Issue by ORBAT** [sanity-io/sanity#1581]
# Problem
If a document field is of some custom type that always has the same default, I have to respecify those defaults in every document that uses it. As an example:

```js
// paid.js
export default {
  name: 'paidContent',
  title: 'Paid content',
  type: 'boolean',
  description: 'Only show this to paying customers',
  options: {
    layout: "checkbox",
  },
}
```

```js
// docs.js
export const someDocument {
  name: 'someDocument',
  fields: [
    // buncha fields here
    {
      name: 'paidContent',
      type: 'paidContent'
    },
  ],

  initialValue: {
    paidContent: true,
  },
}

export const anotherDocument {
  name: 'anotherDocument',
  fields: [
    // buncha fields here
    {
      name: 'paidContent',
      type: 'paidContent',
    }
  ],

  initialValue: {
    paidContent: true,
  },
}

```

# Proposed solution

```js
/// paid.js
export const paidContent =  {
  name: 'paidContent',
  title: 'Paid content',
  type: 'boolean',
  description: 'Only show this to paying customers',
  options: {
    layout: "checkbox",
  },

  initialValue: {
    paidContent: true,
  },
};
```

```js
// docs.js
export const someDocument {
  name: 'someDocument',
  fields: [
    // buncha fields here
    {
      name: 'paidContent',
      type: 'paidContent',
    }
  ]
}

export const anotherDocument {
  name: 'anotherDocument',
  fields: [
    // buncha fields here
    {
      name: 'paidContent',
      type: 'paidContent',
    }
  ]
}

```

Both `someDocument` and `anotherDocument` would have the initial value `{ paidContent: true }`.

# Alternate solutions

E.g. Exporting the initial value, and then importing & merging it with document initial values. More brittle, requires users of the type to do it manually

**Comment by kmelve** [comment]
Thanks for submitting this! It's obviously a feature that might be handy. We have no ETA on it as of yet, but feel free to follow this issue.

**Comment by matthewharwood** [comment]
@kmelve pretty please?

**Comment by jasonbarone** [comment]
I like the proposed solution from @ORBAT. I would also add that a nice addition would be to ensure that the page document can override the default that is set in the type initialValue too.

**Comment by brentrobbins** [comment]
I also look forward to a solution where we can add Initial values for objects.

**Comment by simonhrogers** [comment]
I would also love this to be resolved!

**Comment by MaximilianLloyd** [comment]
This is all I want for Christmas

**Comment by sisselfladby** [comment]
+1 on initial values for objects as well 🙏

**Comment by TimmyR** [comment]
i++ 
And is there a temp workaround for this? And is there any news on the ETA by now?

**Comment by johnnypea** [comment]
I would like to be able to use an initial values for objects as well.

**Comment by remisture** [comment]
+1 Yes, please :)

**Comment by christopherafbjur** [comment]
+1 Looking forward to this!!

**Comment by spacedawwwg** [comment]
+1 for initial values for objects, please.

**Comment by fabifors** [comment]
+1 for initial / default values for objects 🙏

**Comment by djavco** [comment]
+1 for would be very nice to set initial values for objects

**Comment by intensr** [comment]
+1

**Comment by rastko909** [comment]
This would really be awesome! Are there any workarounds?

**Comment by benjaminsehl** [comment]
+1 on this for sure, major pain point

**Comment by nzhenry** [comment]
+1 for this. Any idea where this is on the roadmap?

**Comment by joshfarrant** [comment]
@kmelve Would you accept a PR for this? On the surface it seems fairly straight-forward, or are you expecting it to be more complex?

**Comment by joshfarrant** [comment]
Or @rexxars equally 🙂

^^^

**Comment by bryangruneberg** [comment]
Given how great everything else is, I was super surprised to see default values / initial values aren't possible for objects. Would be great to have a solution to this!

**Comment by kmelve** [comment]
Hi folks! We agree that this feature makes sense. Offering a PR is super generous @joshfarrant, but we will do this one internally as it's tied to code paths that we want to clean up a bit. It's on the prioritized backlog though!

**Comment by netmet12** [comment]
Any updates here? We really need this! Are you sure you don't want to accept a PR?

**Comment by evenwestvang** [comment]
Just wanted to let you know that we have a PR shaping up for this issue. Thank you for your patience.

**Comment by bjoerge** [comment]
If you are keen on trying out this feature before we ship it officially, we just cut a canary release from PR #2444. You can install it in your Studio by running:

```sh
sanity upgrade --tag=canary --save-exact
```

It should work very much like the solution initially outlined by @ORBAT. You can see a full example of a document that uses different variations of initial values here: https://github.com/sanity-io/sanity/blob/d750f0a9e2c51984d5349edc67987ceb04e37413/examples/test-studio/schemas/initialValuesTest.js

**Comment by bjoerge** [comment]
Initial value support for any type has shipped in [v2.10.0](https://github.com/sanity-io/sanity/releases/tag/v2.9.1). Thank you all for your patience and valuable input!

**Comment by spacedawwwg** [comment]
@bjoerge literally came to check for an update on this! nice surprise :)

**Comment by fabien** [comment]
@bjoerge this is a great addition!

However, it appears that the `initialValue` function does not receive any useful arguments.

For a newly created document, this could pass in the document as created using **Initial Value Templates**.

For nested objects (or arrays of objects) this is much more useful, since you'd be able to infer what to add there based on the main document. For example, when adding to a list of attributes (array of objects), you'd be able to lookup the `initialValue` of each item based off the specific document type or other document properties.

In short, for consistency with the other functions in Sanity (custom validation, uniqueness etc.) it would make sense to provide a `context` argument to `initialValue`.
